stdlib/public/core/Diffing.swift

```diff
@@ -0,0 +1,782 @@
+    let pathEnd = (a.endIndex, b.endIndex)
+
+    // `endpoints.storage` may need space for an additional element in order
+    // to store the difference path when `d == 1`.
+    //
+    // `endpoints.storage` has `(d + 1) * (d + 2) / 2` elements stored,
+    // but a difference path requires up to `2 + d * 2` elements[^1].
+    //
+    // If `d == 1`:
+    //
+    //   (1 + 1) * (1 + 2) / 2 < 2 + 1 * 2
+    //             3 < 4
+    //
+    // `d == 1` is the only special case because:
+    //
+    // - It's a precondition that `d > 0`.
+    // - Once `d >= 2` `endpoints.storage` will have sufficient space:
+    //
+    //     (d + 1) * (d + 2) / 2 = 2 + d * 2
+    //    d * d - d - 2 = 0
+    //     (d - 2) * (d + 1) = 0
+    //    d = 2; d = -1
+    //
+    // [1]: An endpoint for every remove, insert, and match segment. (Recall
+    // *d* is the minimal number of inserted and removed elements). If there
+    // are no consecutive removes or inserts and every remove or insert is
+    // sandwiched between matches, the path will need `2 + d * 2` elements.
+    assert(d > 0, "Must be at least one difference between `a` and `b`")
+    if d == 1 {
+      endpoints.storage.append(pathEnd)
+    }
+
+    var i = endpoints.storage.endIndex - 1
+    // `isInsertion` tracks whether the element at `endpoints.storage[i]`
+    // is an insertion (`true`), a removal (`false`), or a match (`nil`).
+    var isInsertion: Bool? = nil
+    var k = delta
+    endpoints.storage[i] = pathEnd
+    for d in (1...d).reversed() {
+      if k == -d || (k != d && self[d - 1, k - 1].x < self[d - 1, k + 1].x) {
+        let (x, y) = self[d - 1, k + 1]
+
+        // There was match before this insert, so add a segment.
+        if x != endpoints.storage[i].x {
+          i -= 1; endpoints.storage[i] = (x, b.index(after: y))
+          isInsertion = nil
+        }
+
+        // If the previous segment is also an insert, overwrite it.
+        if isInsertion != .some(true) { i -= 1 }
+        endpoints.storage[i] = (x, y)
+
+        isInsertion = true
+        k += 1
+      } else {
+        let (x, y) = self[d - 1, k - 1]
+
+        // There was a match before this remove, so add a segment.
+        if y != endpoints.storage[i].y {
+          i -= 1; endpoints.storage[i] = (a.index(after: x), y)
+          isInsertion = nil
+        }
+
+        // If the previous segment is also a remove, overwrite it.
+        if isInsertion != .some(false) { i -= 1 }
+        endpoints.storage[i] = (x, y)
+
+        isInsertion = false
+        k -= 1
+      }
+    }
+
+    if pathStart != endpoints.storage[i] {
+      i -= 1; endpoints.storage[i] = pathStart
+    }
+
+    let pathStorage = endpoints.storage
+    endpoints.storage = []
+    return CollectionChanges(pathStorage: pathStorage, pathStartIndex: i)
+  }
+}
+
+/// An index that counts its offset from the start of its collection.
+struct CountingIndex<Base : Comparable> : Equatable {
```

Comment:
This is implicitly internal, so it should have an underscore. (Maybe it could also be made private.)

```suggestion
internal struct _CountingIndex<Base : Comparable> : Equatable {
```